### PR TITLE
[inductor] add option to autotune at fusion time for mm

### PR DESCRIPTION
**Pull request by warrendeng**
For epilogue fusion with max-autotuning enabled, we first autotune gemms at lowering time to get an initial selection of configs before benchmarking them with fused epilogue. Because we autotune prior to fusion benchmarking, we potentially miss out on configs that are optimal for fusion. 

In this PR, we add an option to move autotuning from lowering to fusion time.

In this implementation, we skip compilation and benchmarking of unfused triton configs in favor of benchmarking all possible fused triton configs. In case where fused configs are slower than Aten, we all back to benchmarking unfused configs. 

This has an advantage over maxing out max_epilogue_benchmarked_choices since we save on autotuning time in cases where fused configs are faster than Aten.

We ran tests on a simple mat mul + relu script and saw compile time improvements over maxing out max_epilogue_benchmarked_choices

image

We also ran benchmarks on OSS HuggingFace suite and saw speedup for some models vs default max-autotune and vs max-autotune + increasing max_epilogue_benchmarked_choices. In these benchmarks, we saw roughly similar compilation time between our new path and prod path. Results are shown below 

image
image
image

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @PaulZhang12

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -4310,6 +4310,80 @@ def log_fusion(ms_fused: float, ms1: float, ms2: float) -> None:
...
+            # TLDR: We fuse all configs and benchmark them fused, comparing against the extern (cuBLAS) baseline. We skip unfused benchmarking
+            if config.autotune_gemm_at_epilogue_fusion_time and epilogue_fusion:
+                # TODO: move CantSplit import to top of file
+                from torch._inductor.codegen.simd import CantSplit
+
+                # 1. Get only extern/cuBLAS baseline timing.
+                #    extern_only=True skips unfused Triton autotuning entirely.
+                choice_timings = multi_node.choice_timings(extern_only=True)
+                extern_time = (
+                    min(choice_timings.values()) if choice_timings else float("inf")
```
Comment (PaulZhang12) [inline torch/_inductor/scheduler.py]: Call autotuning logic itself here, maybe a bit cleaner and less invasive

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -4310,6 +4310,80 @@ def log_fusion(ms_fused: float, ms1: float, ms2: float) -> None:
...
+                # 2. Get epilogue (node2) runtime estimate
+                ms2 = node2._get_estimated_runtime()
+
+                # 3. Compile ALL Triton choices with fusion (no
+                #    max_epilogue_benchmarked_choices limit)
+                future_choices: list[tuple[Any, LambdaFuture | None, ModuleType]] = []
+                for choice in multi_node.choices:
+                    if not isinstance(choice, TritonTemplateCallerBase):
+                        continue
+                    with multi_node.swap_as_triton_caller(choice):
```
Comment (PaulZhang12) [inline torch/_inductor/scheduler.py]: Why is this needed?

File: torch/_inductor/scheduler.py
Comment (warrendeng) [inline torch/_inductor/scheduler.py:4458]: moved this to a shared function _benchmark_fused_choices

File: torch/_inductor/scheduler.py (hunk 2 above)
Comment (warrendeng) [reply torch/_inductor/scheduler.py]: in this loop we iterate through all of the triton configs and compile them

we need this with statement or else compile_kernel will not know which config to use inside of node_list_fused

this pattern appears a lot in the rest of this file (for example https://github.com/pytorch/pytorch/blob/a455f65371f0b9ff51d4397b12a479ea3f8dcd5b/torch/_inductor/scheduler.py#L4517)

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -4366,6 +4382,77 @@ def log_fusion(ms_fused: float, ms1: float, ms2: float) -> None:
...
+                        try:
+                            future_choices.append(
+                                (
+                                    choice,
+                                    *self.compile_kernel(node_list_fused),
+                                )
+                            )
+                        except CantSplit:
```
Comment (PaulZhang12) [inline torch/_inductor/scheduler.py]: What is `CantSplit`?

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -4366,6 +4382,77 @@ def log_fusion(ms_fused: float, ms1: float, ms2: float) -> None:
             if self._has_layout_conflict_for_template(multi_node):
                 return FusionResult.fuse(False)
 
+            # Autotune all Triton configs at fusion time: benchmark each
+            # config fused with the epilogue, comparing against extern baseline.
+            if config.autotune_gemm_at_epilogue_fusion_time and epilogue_fusion:
```
Comment (PaulZhang12) [inline torch/_inductor/scheduler.py]: Is it possible to move most/if not all of this logic into a different helper?

File: torch/_inductor/scheduler.py [hunk 5]
```diff
@@ -4366,6 +4382,77 @@ def log_fusion(ms_fused: float, ms1: float, ms2: float) -> None:
...
+                            )
+                        except CantSplit:
+                            continue
+
+                if not future_choices:
+                    return FusionResult.fuse(False)
+
+                def benchmark_fused_autotune() -> bool:
+                    assert isinstance(multi_node, ir.MultiTemplateBuffer)
+                    min_ms_fused, ms_fused_choice, new_timings = (
```
Comment (PaulZhang12) [inline torch/_inductor/scheduler.py]: We probably eventually want some heuristics here as well, like some sort of smarter search, where say if the current choice can't do the fusion because it runs out of shared memory, ensure the total shared memory from block sizes + num_stages of future choices to at least be smaller than the current one

File: torch/_inductor/scheduler.py (hunk 3 above)
Comment (warrendeng) [reply torch/_inductor/scheduler.py]: this is also copied from the prod logic https://github.com/pytorch/pytorch/pull/178338/changes#diff-d2539c9c8dc6a3d7e457767a880612e96d3c85752a77ead49a9e4e00a3e4c3c7L4465

the exception happens when the fused kernel's iteration ranges aren't compatible with the config's tiling and we end up skipping that config if it happens

File: torch/_inductor/scheduler.py (hunk 4 above)
Comment (warrendeng) [reply torch/_inductor/scheduler.py]: migrated to a helper

File: test/inductor/test_benchmark_fusion.py [hunk 6]
```diff
@@ -345,6 +345,34 @@ def fn(a: torch.Tensor, b: torch.Tensor, c: torch.Tensor):
 
             torch._dynamo.reset()
 
+        @fresh_cache()
```
Comment (warrendeng) [inline test/inductor/test_benchmark_fusion.py:348]: added test for new path and verified it passes using 

`TORCHINDUCTOR_COMPILE_THREADS=1 python test/inductor/test_benchmark_fusion.py BenchmarkMultiTemplateFusionGpuTest.test_autotune_fused_epilogue` cc @PaulZhang12

File: torch/_inductor/scheduler.py [hunk 7]
```diff
@@ -4108,6 +4108,119 @@ def benchmark_codegened_module(
...
+                continue
+            with multi_node.swap_as_triton_caller(choice):
+                try:
+                    future_choices.append(
+                        (
+                            choice,
+                            *self.compile_kernel(node_list_fused),
+                        )
+                    )
+                except CantSplit:
```
Comment (PaulZhang12) [inline torch/_inductor/scheduler.py:4252]: Sorry can you explain the `CantSplit` change here? I think I asked it before, not sure if I missed

File: test/inductor/test_benchmark_fusion.py [hunk 8]
```diff
@@ -345,6 +345,37 @@ def fn(a: torch.Tensor, b: torch.Tensor, c: torch.Tensor):
...
+            from unittest.mock import patch as mock_patch
+
+            def foo(m, inp):
+                return torch.nn.functional.relu(m(inp))
+
+            m = torch.nn.Linear(256, 256, bias=True).half().to(GPU_TYPE)
+            inp = torch.rand([256, 256]).half().to(GPU_TYPE)
+
+            original = Scheduler._autotune_fused_epilogue
+            called = [False]
```
Comment (PaulZhang12) [inline test/inductor/test_benchmark_fusion.py]: Nit: this test can maybe be made a bit cleaner? No need for a list here, just nonlocal variable is fine

File: torch/_inductor/config.py [hunk 9]
```diff
@@ -506,6 +506,12 @@ def prologue_fusion_enabled() -> bool:
 # enable slow autotuning passes to select gemm algorithms
 max_autotune_gemm = os.environ.get("TORCHINDUCTOR_MAX_AUTOTUNE_GEMM") == "1"
 
+# Autotune gemm configs at epilogue fusion time, skipping unfused autotuning
+autotune_gemm_at_epilogue_fusion_time = (
+    os.environ.get("TORCHINDUCTOR_AUTOTUNE_GEMM_AT_EPILOGUE_FUSION_TIME") == "1"
```
Comment (PaulZhang12) [inline torch/_inductor/config.py]: nit: maybe rename to without epilogue, as we want to keep it general fusion time

File: torch/_inductor/scheduler.py (hunk 7 above)
Comment (warrendeng) [reply torch/_inductor/scheduler.py:4252]: oops yeah you commented it, accidentally resolved the comment here https://github.com/pytorch/pytorch/pull/178338#discussion_r3011891222

**Review by eellison — COMMENTED** [review]
Looks good, but, I don't think this is preventing us compiling the initial configs during lowering. Is that intentional ? Otherwise what's the point. Other than just increasing `max_epilogue_benchmarked_choices`

File: torch/_inductor/config.py (hunk 9 above)
Comment (warrendeng) [reply torch/_inductor/config.py]: thanks, renamed

**Comment by warrendeng** [comment]
Thanks for taking a look @eellison !

> I don't think this is preventing us compiling the initial configs during lowering. Is that intentional ?

For the new path we actually do skip compilation and benchmarking of the unfused configs since we no longer call choice_timings where the deferred compilation happens (https://github.com/pytorch/pytorch/pull/178338/changes#diff-d2539c9c8dc6a3d7e457767a880612e96d3c85752a77ead49a9e4e00a3e4c3c7L4393)

The overall flow should be:
- We benchmark Aten/cuBLAS as a baseline to compare fusion performance against
- We benchmark all fused Triton configs 
- If a fused Triton config wins, we use it. Otherwise we don't fuse, and the node falls through to finalize_multi_template_buffers where unfused autotuning happens as normal

So in the case where the fused config is not picked, yes this would be equivalent to maxing out `max_epilogue_benchmarked_choices`, but if it is picked we save from having to do unfused autotuning

**Comment by eellison** [comment]
Choice timings should just trigger the benchmarking. but the precompilation should fire previously. can you verify this ?

**Comment by eellison** [comment]
I guess for kernels which dont have fusion opportunities, we might still want to autotune them....

do we have benchmarking data of improvements with this ? the advantage of benchmarking the base mms is that we can do a better job triggering them early.

Id be curious about perf/compile time of this compared to increasing `max_epilogue_benchmarked_choices`.

**Comment by eellison** [comment]
@warrendeng can you try this on training as well ? I believe in training epilogue fusion is less successful. if we have an idea of which matmuls we expect to epilogue fuse or not, we could lower them differently. 

Also, do we really need to try all 20 ? 

Since this is off by default i'm fine landing. just if we want to try to turn on, maybe we resolve/think about some of these cases.
